src/utils/cache.ts

```diff
@@ -0,0 +1,45 @@
+import { stat } from './fs';
+import { HookCache, Plugin } from './rollup/types';
+
+export function cacheResolveIdHook(cache: HookCache, plugin: Plugin) {
+	return function cachedResolveId(id: string) {
+		const key = `resolve|${plugin.name || '(anonymous plugin)'}|${plugin.cacheKey}|${id}`;
+		return key in cache
+			? Promise.resolve(cache[key])
+			: Promise.resolve()
+					.then(() => plugin.resolveId.call(this, id))
+					.then(value => (cache[key] = value));
+	};
+}
+
+export function cacheLoadHook(cache: HookCache, plugin: Plugin) {
+	return function cachedLoad(id: string) {
+		// An inherently side-effectful part of `.load` is the filesystem.
+		// We want to invalidate the cache if the file has changed.
+		// A good cheap metric for this is mtime and size. We combine these into the
+		// cache key so that files with different mtime/size invalidate the cache
+		return stat(id)
```

Comment:
We have to be weary of tying the implementation to the file system. Rollup runs equally in a browser environment, with the default load hook only applying if no other plugin is present. Also module Ids in Rollup don't strictly have to be file paths, so we need to be aware of this as well.

For this reason, my initial proposal was, rather than specifying an exact cache key implementation like you have done here, leave this exact part to the plugins themselves, offering a lower level primitive that doesn't tie to the file system.

Reply:
The simple answer here is that `stat` is caught, so any failure to read from disk (including the error thrown from the stubbed version) should be caught and the `[0, 0]` tuple returned.

The issue with allowing plugins to generate their own keys is it leaves a lot of responsibility on them to get it right. This - to me - could be problematic as plugins are often not as well tested as core, and missing pieces of state like the filesystem can surface.

If we were to allow plugins to generate their own keys I think we'd need to define a way for plugins to hit the disk with an object we can introspect or otherwise control, so that the load plugin doesn't just return file contents, but returns a file object (think something like [vynil fs](https://github.com/gulpjs/vinyl-fs)). We can then use the file object to decide how it can be cached (e.g. if the file object represents a file on disk, then stat it, if it is something else then apply other rules). If the only way `load` can operate is to return a `File` object then that changes the dynamic and allows us to do alot (e.g. utilising a `File.hasChanged() === true?`)

Reply:
So here's how I'd imagine a plugin implementing the load hook with a `cacheKey`:

```js
async load (id, cacheKey) {
  const [mtime, size] = cacheKey.split('|');
  const stats = await new Promise((resolve, reject) => 
    fs.stat(id, (stats, err) => err ? reject(err) : resolve(stats)
  );
  if (mtime === stats.mtime && size === stats.size)
    return true;
  // ...usual load routine
}
```

if we document this well, it should be clear to plugins how to use this, and most will likely just copy-paste the snippet anyway.

The benefit here is that plugins then have the flexibility to extend and modify this behaviour based on their exact load method, and I think that will be clear given the above snippet given to them.

We do have the plugin lists, so we could even PR these ourselves once we can agree on the API format.

I'm completely open to API alternatives here as well too.